### ISSUE TITLE
False positive for `Lint/RedundantWithIndex` when calling with_index without each beforehand

### ISSUE THREAD

**Issue by Earlopain** [rubocop/rubocop#12769]
This cop is marked as safe, however some code can result in offenses that when autocorrected break the code.

--------

## Expected behavior

No offense for the following:

```rb
ary = [1, 2, 3, 4].each
ary.with_index { |v| v }
```

## Actual behavior

Autocorrection to the following:

```rb
ary = [1,2,3,4].each
ary { |v| v }
```

This will raise when executed: ``undefined method `ary'``. It will also promt `Lint/UselessAssignment` on the first line since that variable is now unused.

## RuboCop version

```
$ [bundle exec] rubocop -V
1.62.0 (using Parser 3.3.0.4, rubocop-ast 1.31.1, running on ruby 3.3.0) [x86_64-linux]
  - rubocop-performance 1.20.2
  - rubocop-rake 0.6.0
  - rubocop-rspec 2.27.1
```
